### PR TITLE
Maximize file sizes produced by each materialize task

### PR DESCRIPTION
**Pull request by rkenmi**
Issue: #65 

Create read-optimized number of output files in materialize while retaining minimal heap memory usage.

Example compaction job with `records_per_output_file = 100_000`, each debug output = one materialize task:
```csv
  | 2023-02-08T17:52:29.725-08:00 | 2023-02-09 01:52:29,530 DEBUG materialize.py:226 -- 3 files written with records: [700000, 100000, 53352]
  | 2023-02-08T17:52:32.038-08:00 | 2023-02-09 01:52:32,021 DEBUG materialize.py:226 -- 3 files written with records: [700000, 200000, 58439]
  | 2023-02-08T17:52:43.315-08:00 | 2023-02-09 01:52:43,143 DEBUG materialize.py:226 -- 3 files written with records: [300000, 100000, 33628]
  | 2023-02-08T17:52:43.342-08:00 | 2023-02-09 01:52:43,160 DEBUG materialize.py:226 -- 3 files written with records: [700000, 100000, 12591]
  | 2023-02-08T17:52:43.757-08:00 | 2023-02-09 01:52:43,508 DEBUG materialize.py:226 -- 3 files written with records: [700000, 100000, 59279]
  | 2023-02-08T17:52:47.608-08:00 | 2023-02-09 01:52:47,587 DEBUG materialize.py:226 -- 3 files written with records: [200000, 100000, 41779]
```

### CODE REVIEW

File: deltacat/compute/compactor/steps/materialize.py [hunk 1]
```diff
@@ -154,8 +149,7 @@ def _materialize(
             src_file_idx_np.item(),
             file_reader_kwargs_provider=read_kwargs_provider,
         )
-        record_count = len(pa_table)
-        mask_pylist = list(repeat(False, record_count))
+        mask_pylist = list(repeat(False, len(pa_table)))
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:152]: We don't need to create a mask pylist after my latest changes. We could directly pass pyarrow array here to `take` indices that we need.

File: deltacat/compute/compactor/steps/materialize.py [hunk 2]
```diff
@@ -181,27 +175,52 @@ def _materialize(
             [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
         )
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:176]: Sorting must be done after merging all the slices that go into a single file. However, we would not need this sorting if we are not predicting the indices using record counts.

File: deltacat/compute/compactor/steps/materialize.py [hunk 3]
```diff
@@ -181,27 +175,52 @@ def _materialize(
...
 
-        # Write manifests up to max_records_per_output_file
-        # TODO(raghumdani): Write exactly the same number of records into each file to
-        # produce a read-optimized view of the tables.
-        if compacted_tables and \
-                total_record_count + record_count > max_records_per_output_file:
-            materialized_results.append(_materialize(compacted_tables, total_record_count))
+        if remainder_tables:
+            prev_remainder_record_count = sum([len(tbl) for tbl in remainder_tables])
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:179]: This may become inefficient if we end up iterating through all the remainder tables especially if there are lot of small tables and our max_records_per_output_file is too huge. Basically, increasing the complexity from O(N) to skim through this to O(N^2).

File: deltacat/compute/compactor/steps/materialize.py [hunk 4]
```diff
@@ -181,27 +175,52 @@ def _materialize(
...
+                f" to be fitted to {max_records_per_output_file}, but got {remainder_records} instead."
+            materialized_results.append(_materialize(remainder_tables))
             # Free up written tables in memory
-            compacted_tables.clear()
-            total_record_count = 0
+            remainder_tables.clear()
+
+            # Update the starting record offset of the current file
+            if new_file_record_offset < len(pa_table):
+                pa_table = pa_table.slice(offset=new_file_record_offset)
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:203]: Have you tested for the case where this table turns out to be empty?

File: deltacat/utils/pyarrow.py [hunk 5]
```diff
@@ -133,6 +133,24 @@ def slice_table(
     return tables
 
 
+def slice_table_generator(
+        table: pa.Table,
+        max_len: int) -> Generator[pa.Table, None, None]:
+    """
```
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:139]: You could document this public method, and mention this wouldn't yield any results when the table is empty.

File: deltacat/utils/pyarrow.py [hunk 6]
```diff
@@ -133,6 +133,24 @@ def slice_table(
...
+    Generator for creating 0-copy table slices.
+    """
+    offset = 0
+    records_remaining = len(table)
+    while records_remaining > 0:
```
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:144]: When the max records per file is say 100 (N) but your table has 100000 (M) records, this would yield 1000 smaller pyarrow tables. The 1000 compute iterations could be saved if you just split it into two parts one of size `floor(M/N)*N` and second of `M%N`.

File: deltacat/compute/compactor/steps/materialize.py [hunk 7]
```diff
@@ -154,8 +149,7 @@ def _materialize(
             src_file_idx_np.item(),
             file_reader_kwargs_provider=read_kwargs_provider,
         )
-        record_count = len(pa_table)
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:157]: This was a bug initially which counted record counts of original `pa_table` and not the filtered table. We may need to setup unit test and integration to catch these kind of issues going forward.

**Review by raghumdani — COMMENTED** [review]
Overall, as advertised. However, you may need to do some more optimizations to merge.

File: deltacat/utils/pyarrow.py (hunk 6 above)
Comment (raghumdani) [reply deltacat/utils/pyarrow.py:151]: The idea here is to save the 1000 compute iterations everytime we add a new entry to remainder table.

Comment (rkenmi) [reply deltacat/utils/pyarrow.py:151]: The slicing only occurs on `stage_delta` if the table being passed in has greater record counts than `max_records_per_output_file`. So by effectively passing in a concatenated table with a guaranteed size of `max_records_per_output_file` (or less than, for the final remainder), we can guarantee that each concatenated table would result in one file written to S3 (and not have any further splitting).

Basically, pre-splitting to `[100, 100, 100, ..., 100, 1]` or just bundling upfront to `[100_000, 1]` doesn't really change the fact that we compute 1000 iterations. It might be a bit nuanced behavior, but pre-splitting is convenient here since we are also slicing and merging table records based on the offset of previous remainders.

Comment (raghumdani) [reply deltacat/utils/pyarrow.py:151]: Correct me if I am wrong, don't we end up calling data catalog multiple times to create new "in-progress" delta for each file instead of calling data catalog once for all the files uploaded during a single `stage_delta` call.

File: deltacat/compute/compactor/steps/materialize.py [hunk 8]
```diff
@@ -45,16 +47,20 @@ def materialize(
...
+        logger.debug(f"Concatenating {len(compacted_tables)} compacted tables"
+                     f" with total size: {compacted_tables_size} bytes")
         compacted_table = pa.concat_tables(compacted_tables)
+        compacted_table = compacted_table.take(
+            pc.sort_indices(compacted_table, sort_keys=PA_SORT_KEYS),
+        )
+        compacted_table = compacted_table.drop(
+            [sc._DEDUPE_TASK_IDX_COLUMN_NAME]
+        )
+
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:63]: To be precise, this sorting can be removed as we are not predicting the indices during dedupe.

File: deltacat/utils/pyarrow.py (hunk 6 above)
Comment (rkenmi) [reply deltacat/utils/pyarrow.py:151]: Currently, the data catalog API internally slices a given table into groups of `max_records_per_output_file` tables, and performs a single S3 call per table.

https://github.com/ray-project/deltacat/blob/d4b307cae12c8d8a6f6ccbc1fe6169e8a3fd20e5/deltacat/aws/s3u.py#L281

Comment (rkenmi) [reply deltacat/utils/pyarrow.py]: I've since revised my changes to do post-splitting instead, since the data catalog API call itself could be expensive and it would be good to reduce the number of these calls as much as we can (we'll still have to live with more than a single stage_delta call per materialize task, if we eagerly materialize).

File: deltacat/compute/compactor/steps/materialize.py [hunk 9]
```diff
@@ -92,7 +94,10 @@ def _materialize(
         logger.info(f"Materialize result: {materialize_result}")
         return materialize_result
 
-    logger.info(f"Starting materialize task...")
+    # TODO (ricmiyam): Update this for full implementation of https://github.com/ray-project/deltacat/issues/62
+    materialize_task_id = get_runtime_context().get_task_id()
+    logger.info(f"Starting materialize task {materialize_task_id}...")
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:99]: Minor: Also what could be helpful here is to print materialize bucket index.

File: deltacat/compute/compactor/steps/materialize.py [hunk 10]
```diff
@@ -164,49 +172,66 @@ def _materialize(
...
-        # TODO(raghumdani): Write exactly the same number of records into each file to
-        # produce a read-optimized view of the tables.
-        if compacted_tables and \
-                total_record_count + record_count > max_records_per_output_file:
-            materialized_results.append(_materialize(compacted_tables, total_record_count))
+            # Chop off records that are already written to disk
+            pa_table = pa_table.slice(offset=records_to_fit)
+
+        record_count = len(pa_table)
+        assert record_count > 0, f"Expected non-empty compacted PyArrow table"
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:195]: Would this always be true? What happens when `remainder_record_count + len(pa_table)` is equal to `max_records_per_output_file` ?

File: deltacat/compute/compactor/steps/materialize.py [hunk 11]
```diff
@@ -164,49 +172,66 @@ def _materialize(
...
+            materialized_results.append(_materialize(remainder_tables))
             # Free up written tables in memory
-            compacted_tables.clear()
-            total_record_count = 0
+            remainder_tables.clear()
+            remainder_record_count = 0
 
-        total_record_count += record_count
-        compacted_tables.append(pa_table)
+        # Add any leftovers (less than max_records_per_output_file) to the remainder list after packing
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:211]: Instead of adding comments what the code next to it is doing (which is fairly easy to understand), what you may add is a doc to this method and explain `why` part i.e., why does it have to add left overs to remainder table.

File: deltacat/compute/compactor/steps/materialize.py (hunk 10 above)
Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py]: No, thanks for catching (this is why we need unit tests) 🙂

File: deltacat/compute/compactor/steps/materialize.py (hunk 11 above)
Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py:210]: How about a separate issue altogether for adding docstrings to each of the compaction steps?

File: deltacat/compute/compactor/steps/materialize.py [hunk 12]
```diff
@@ -164,49 +172,66 @@ def _materialize(
...
+            if remainder_record_count + len(pa_table) < max_records_per_output_file:
+                remainder_tables.append(pa_table)
+                remainder_record_count += len(pa_table)
+                continue
+
+            # 'max_records_per_output_file' will be exceeded with the
+            #  current file's records + previous remainder records.
+            records_to_fit = max_records_per_output_file - remainder_record_count
+            fitted_table = pa_table.slice(length=records_to_fit)
+            remainder_tables.append(fitted_table)
```
Comment (raghumdani) [inline deltacat/compute/compactor/steps/materialize.py:189]: This version looks good. However, I still feel it could be further modularized to make it easy to understand. Remember, we also have to add indices to these records as and when they are getting materialized to support primary key index building in future. In the current state, it may be difficult to make those modifications. I have proposed some changes below. Let me know if that makes sense:

```python
# This class handles splitting the tables and letting us know when it's ready to materialize
class BatchedTables(List[pa.Table]):
    def __init__(self, max_records_per_table_multiple):
        self.__init__()
        self._tables_ready = []
        self._tables_leftover = []
        self._tables_leftover_len = 0
        self._tables_ready_len = 0
    def append(self, table: pa.Table):
        # split into two in the multiples of max_records and left over and append to
        # appropriate lists created above
        if (_tables_leftover_len > max_records):
            # time to split move some of them to _ready_tables

    def get_ready_tables() -> List:
        return self._tables_ready

    def clear_ready() -> List:
        self._tables_ready.clear()

    def get_remainder_tables() -> List:
        return self._tables_leftover
  
# In your materialize code

batched_tables = BatchedTables(max_records_per_file)

for table in tables:
    batched_tables.append(table)
    if (len(batched_tables.get_ready_tables()) > 0):
        _materialize(batched_tables.get_ready_tables())
        batched_tables.clear_ready_tables()

if (batched_tables.get_ready_tables()):
    _materialize(batched_tables.get_ready_tables())
    batched_tables.clear_ready_tables()

if (batched_tables.get_remainder_tables()):
    _materialize(batched_tables.get_remainder_tables())

```

File: deltacat/utils/pyarrow.py [hunk 13]
```diff
@@ -18,7 +18,7 @@
 from deltacat.utils.common import ReadKwargsProvider, ContentTypeKwargsProvider
 from deltacat.utils.performance import timed_invocation
 
-from typing import Any, Callable, Dict, List, Optional, Iterable, Union
+from typing import Any, Callable, Dict, List, Optional, Iterable, Union, Generator
```
Comment (raghumdani) [inline deltacat/utils/pyarrow.py:21]: Is this needed now?

**Review by raghumdani — COMMENTED** [review]
Looks good from performance aspect. Please try to address some minor comments related to readability.

File: deltacat/compute/compactor/steps/materialize.py (hunk 11 above)
Comment (raghumdani) [reply deltacat/compute/compactor/steps/materialize.py:210]: Makes sense.

File: deltacat/compute/compactor/steps/materialize.py (hunk 12 above)
Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py:189]: I like the general idea. I think there could be an unintended performance hit here when doing the split to move `self._tables_leftover` over to `self._ready_tables` too frequently or across too many tables?

Comment (raghumdani) [reply deltacat/compute/compactor/steps/materialize.py]: I think we can mitigate that by using `double ended queue` if required as we have to maintain table ordering.

Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py]: In the class you provided, I was more so referring to how we would need to concatenate leftover tables and split them, each time we need to shift `self._tables_leftover` over to `self._ready_tables`, since we maintain two lists of tables now. What if max records is `4000`, and each table added has 3999 records?

Comment (raghumdani) [reply deltacat/compute/compactor/steps/materialize.py]: Doesn't the current implementation also require you to split pyarrow table everytime in the above case? As far as shifting the table goes, you will only have shift a particular table only once, so it will not affect performance. You can optimally shift if you use `double ended queues` instead of lists.

Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py]: The current impl. splits once per table, and enforce the remainder to always be less than max records for the next iteration. In this multi-list (or deque) approach we will split once per table *and* also keep track of the leftovers, triggering a shift if it exceeds max records. 

For example
```
tables = [7999, 7999, 7999, ...]

bt.append(tables[0]) # leftover = [3999], ready = [4000] (one split)
bt.append(tables[1]) # leftover = [3998], ready = [4000, 4000, 4000] (one split + one shift)
bt.append(tables[2]) # leftover = [3997], ready = [4000, 4000, 4000, 4000, 4000] (one split + one shift)
...
```

Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py]: I think the point being is that there can be other interesting ways to have this BatchedTable class, but my question is whether or not this modularization has to be P0 itself? We could have a separate issue/PR to refactor this into a class, and even have class-specific unit tests which will be easier to write vs. writing unit tests for the materialize task as a whole.

Comment (raghumdani) [reply deltacat/compute/compactor/steps/materialize.py]: You may think this is not p0, but without the unit tests a small change could cause data corruption. Also, you will find it hard to unit test this code vs modularized code. That way it will be easier for someone to make changes and add new features.

File: deltacat/tests/utils/test_record_batch_tables.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
+import unittest
+import pyarrow as pa
+
+from deltacat.utils.pyarrow import RecordBatchTables
+
+
+def is_sorted(batched_tables: RecordBatchTables,
```
Comment (raghumdani) [inline deltacat/tests/utils/test_record_batch_tables.py:7]: Minor: these methods are not reusable across different tests, so we could move them to the end of the file and start with `_` (underscore).

**Review by raghumdani — APPROVED** [review]
LGTM!

**Review by pdames — APPROVED** [review]
Looks good to merge. Thanks for the unit tests! Minor refactoring can be addressed either prior to merge or in a follow-up PR.

File: deltacat/compute/compactor/steps/materialize.py [hunk 15]
```diff
@@ -92,7 +96,11 @@ def _materialize(
         logger.info(f"Materialize result: {materialize_result}")
         return materialize_result
 
-    logger.info(f"Starting materialize task...")
+    # TODO (ricmiyam): Update this for full implementation of https://github.com/ray-project/deltacat/issues/62
+    materialize_task_id = get_runtime_context().get_task_id()
```
[the remark this answers is not in the document]
Comment (rkenmi) [reply deltacat/compute/compactor/steps/materialize.py:100]: From my compaction E2E test runs (including hash bucketing, dedupe, materialize), the use of `get_runtime_context()` didn't add any noticeable latency and the time taken for the whole workflow to complete was relatively the same. We also just merged a PR (https://github.com/ray-project/deltacat/pull/85) to inject runtime context into our logs, so I can clean up the runtime_context usages here.
